File: server/game/projectiles.py
"""Projectile tracking for bow arrows."""

from __future__ import annotations

import logging
import math
from typing import TYPE_CHECKING

from server.config import settings
from server.game.damage import apply_damage
from server.game.obstacles import line_intersects_obstacle

if TYPE_CHECKING:
    from server.game.state import BotState, Obstacle, Projectile

logger = logging.getLogger(__name__)

_next_id = 0


def create_projectile(
    owner: BotState, target_pos: tuple[float, float], damage: float
) -> Projectile:
    """Create a new projectile from owner toward target position."""
    from server.game.state import Projectile

    global _next_id
    _next_id += 1

    dx = target_pos[0] - owner.position[0]
    dy = target_pos[1] - owner.position[1]
    dist = math.sqrt(dx * dx + dy * dy)
    if dist == 0:
        direction = (1.0, 0.0)
    else:
        direction = (dx / dist, dy / dist)

    speed = settings.combat.projectile_speed
    max_age = int(settings.combat.projectile_max_age_secs * settings.game.tick_rate)

    return Projectile(
        projectile_id=f"arrow_{_next_id}",
        owner_id=owner.bot_id,
        position=owner.position,
        direction=direction,
        speed=speed,
        damage=damage,
        weapon="bow",
        max_age_ticks=max_age,
    )


def update_projectiles(
    projectiles: list[Projectile],
    bots: dict[str, BotState],
    obstacles: list[Obstacle],
    tick_rate: int,
) -> list[dict]:
    """Move projectiles, check hits, return hit events. Removes expired/hit projectiles."""
    hit_radius = settings.combat.projectile_hit_radius + settings.game.bot_radius
    hit_radius_sq = hit_radius * hit_radius
    dt = 1.0 / tick_rate
    events: list[dict] = []
    to_remove: list[int] = []

    for i, proj in enumerate(projectiles):
        proj.age_ticks += 1
        old_pos = proj.position

        # Move projectile
        move_dist = proj.speed * dt
        new_x = proj.position[0] + proj.direction[0] * move_dist
        new_y = proj.position[1] + proj.direction[1] * move_dist
        proj.position = (new_x, new_y)

        # Check obstacle collision (LOS blocked)
        if line_intersects_obstacle(old_pos[0], old_pos[1], new_x, new_y, obstacles):
            to_remove.append(i)
            continue

        # Check arena bounds
        if new_x < 0 or new_x > settings.game.arena_width or \
           new_y < 0 or new_y > settings.game.arena_height:
            to_remove.append(i)
            continue

        # Check hit against bots
        hit = False
        for bot_id, bot in bots.items():
            if bot_id == proj.owner_id or not bot.is_alive:
                continue
            dx = bot.position[0] - new_x
            dy = bot.position[1] - new_y
            if dx * dx + dy * dy <= hit_radius_sq:
                # Invulnerable targets consume the projectile but take no damage
                if bot.invuln_ticks > 0:
                    to_remove.append(i)
                    hit = True
                    break
                # Use apply_damage for proper stat tracking and kill attribution
                apply_damage(bots, bot_id, proj.damage, proj.owner_id, proj.weapon, events)
                to_remove.append(i)
                hit = True
                break

        if hit:
            continue

        # Check age
        if proj.age_ticks >= proj.max_age_ticks:
            to_remove.append(i)

    # Remove in reverse order
    for i in reversed(to_remove):
        projectiles.pop(i)

    return events

File: go-arena/internal/game/projectiles.go
package game

import (
	"arena-server/internal/config"
	"math"
)

// UpdateProjectiles advances all projectiles, checks for collisions with
// obstacles and bots, applies damage on hit, and removes expired or collided
// projectiles.
func UpdateProjectiles(projectiles *[]Projectile, bots map[string]*BotState, obstacles []Obstacle, arenaEvents *[]ArenaEvent, tickCount int, dt float64) {
	alive := (*projectiles)[:0]

	for i := range *projectiles {
		proj := &(*projectiles)[i]
		prevPos := proj.Position

		// 1. Move the projectile.
		proj.Position = proj.Position.Add(proj.Direction.Scale(proj.Speed * dt))
		proj.AgeTicks++

		// 2. Check max age.
		if proj.AgeTicks >= proj.MaxAge {
			if proj.Weapon == "bow" && arenaEvents != nil {
				*arenaEvents = append(*arenaEvents, buildBowImpactEvent(proj.ID, proj.OwnerID, proj.Color, proj.Position, tickCount, "", proj.Intensity))
			}
			continue // Remove: expired
		}

		// 3. Check obstacle/terrain collision.
		//    Ray-march from previous position to current position so fast
		//    projectiles can never skip over a wall cell.
		if ActiveTerrain != nil {
			curCell := ActiveTerrain.WorldToGrid(proj.Position)

			// Exact voxel traversal of the swept segment so fast projectiles
			// can never skip over a wall cell, even on steep diagonals.
			hitWall := ActiveTerrain.IsBlocked(curCell[0], curCell[1]) ||
				ActiveTerrain.SegmentBlocked(prevPos, proj.Position)
			if hitWall {
				if proj.Weapon == "bow" && arenaEvents != nil {
					*arenaEvents = append(*arenaEvents, buildBowImpactEvent(proj.ID, proj.OwnerID, proj.Color, proj.Position, tickCount, "", proj.Intensity))
				}
				continue // Remove: hit wall
			}
		} else if CollidesWithObstacle(proj.Position.X(), proj.Position.Y(), obstacles, 0.5) != nil {
			if proj.Weapon == "bow" && arenaEvents != nil {
				*arenaEvents = append(*arenaEvents, buildBowImpactEvent(proj.ID, proj.OwnerID, proj.Color, proj.Position, tickCount, "", proj.Intensity))
			}
			continue // Remove: hit obstacle
		}

		// 4. Check bot hits using a swept segment so fast projectiles can still
		// connect without landing in the exact same grid cell. When several
		// bots overlap the segment, hit the EARLIEST one along the flight
		// path — iterating the bots map and taking the first match picked a
		// random victim (map order), sometimes the bot standing behind the
		// one actually blocking the shot.
		projectileHitRadius := proj.HitRadius
		if projectileHitRadius <= 0 {
			projectileHitRadius = config.C.ProjectileHitRadius
		}
		hitRadius := config.C.BotRadius + projectileHitRadius

		var victim *BotState
		bestT := math.MaxFloat64
		for _, bot := range bots {
			if !bot.IsAlive || bot.BotID == proj.OwnerID {
				continue
			}
			dist, t := distanceAndParamToSegment(bot.Position, prevPos, proj.Position)
			if dist > hitRadius {
				continue
			}
			if t < bestT {
				bestT = t
				victim = bot
			}
		}

		if victim != nil {
			owner, ownerOk := bots[proj.OwnerID]

			if victim.InvulnTicks > 0 {
				// Invulnerable: consume projectile, no damage.
				if proj.Weapon == "bow" && arenaEvents != nil {
					*arenaEvents = append(*arenaEvents, buildBowImpactEvent(proj.ID, proj.OwnerID, proj.Color, victim.Position, tickCount, victim.BotID, proj.Intensity))
				}
				continue // Remove: hit a bot
			}

			if ownerOk {
				ApplyDamage(victim, owner, proj.Damage, proj.Weapon, tickCount)
				ApplyAttributedGridKnockback(victim, owner, proj.Position.Sub(proj.Direction.Scale(proj.Speed*dt)), 1, obstacles, proj.Weapon, tickCount)
				owner.RoundShotsHit++
			}
			if proj.Weapon == "bow" && arenaEvents != nil {
				*arenaEvents = append(*arenaEvents, buildBowImpactEvent(proj.ID, proj.OwnerID, proj.Color, victim.Position, tickCount, victim.BotID, proj.Intensity))
			}

			continue // Remove: hit a bot
		}

		// 5. Check out of bounds.
		if proj.Position.X() < 0 || proj.Position.X() > config.C.ArenaWidth ||
			proj.Position.Y() < 0 || proj.Position.Y() > config.C.ArenaHeight {
			if proj.Weapon == "bow" && arenaEvents != nil {
				*arenaEvents = append(*arenaEvents, buildBowImpactEvent(proj.ID, proj.OwnerID, proj.Color, proj.Position, tickCount, "", proj.Intensity))
			}
			continue // Remove: out of arena
		}

		// Projectile survives this tick.
		alive = append(alive, *proj)
	}

	*projectiles = alive
}

func distancePointToSegment(point, segStart, segEnd Vec2) float64 {
	dist, _ := distanceAndParamToSegment(point, segStart, segEnd)
	return dist
}

// distanceAndParamToSegment returns the distance from point to the segment
// and the clamped projection parameter t in [0,1] (how far along the segment
// the closest approach happens — used to order projectile victims).
func distanceAndParamToSegment(point, segStart, segEnd Vec2) (float64, float64) {
	seg := segEnd.Sub(segStart)
	segLenSq := seg.X()*seg.X() + seg.Y()*seg.Y()
	if segLenSq <= 1e-9 {
		return point.DistanceTo(segStart), 0
	}

	toPoint := point.Sub(segStart)
	t := (toPoint.X()*seg.X() + toPoint.Y()*seg.Y()) / segLenSq
	t = math.Max(0, math.Min(1, t))

	closest := segStart.Add(seg.Scale(t))
	return point.DistanceTo(closest), t
}
